### PR TITLE
new GraspPlanning service to replace manipulation_msgs version

### PR DESCRIPTION
**Pull request by Jntzko**
Until now the move group pick action uses the grasp planning service from manipulation_msgs.
However, this doesn't use moveit_msgs::Grasp which has evolved since then.

The new service in moveit_msgs uses its own Grasp.msg and overall aims to replace the old functionality
using moveit types.

### CODE REVIEW

**Comment by v4hn** [comment]
According to the discussion in https://github.com/ros-interactive-manipulation/manipulation_msgs/issues/9
we want to move away from `manipulation_msgs` in MoveIt.
@Jntzko and I propose this as a replacement for `manipulation_msgs/GraspPlanning.srv`.

While this should *replace* the old service in MoveIt kinetic, https://github.com/ros-planning/moveit/pull/264 proposes to make an *additional* interface for the new service definition available in the `MoveGroup(Interface)` too.
This change is ABI compatible and could (imho should :)) be made available in MoveIt indigo too.
Thus we'd like to have the definition in indigo too.

File: srv/GraspPlanning.srv [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Requests that grasp planning be performed for the target
```
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: "target *object*"

File: srv/GraspPlanning.srv [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# returns a list of grasps to be tested and executed
```
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: "*candidate* grasps"

File: srv/GraspPlanning.srv [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+# the planning group used
+string group_name
+
+# the object to be grasped
+CollisionObject target
```
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: how about target_object or collision_object?

File: srv/GraspPlanning.srv [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+# the name that the support surface (e.g. table) has in the collision map
+# can be left empty if no name is available
+string collision_support_surface_name
+
+# an optional list of grasps to be evaluated by the planner
+Grasp[] grasps_to_evaluate
```
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: candidate_grasps?

File: srv/GraspPlanning.srv (hunk 3 above)
Comment (v4hn) [inline srv/GraspPlanning.srv]: The name is taken from the original `manipulation_msgs/GraspPlanning.srv`.
I prefer the short name here.

File: srv/GraspPlanning.srv (hunk 4 above)
Comment (v4hn) [inline srv/GraspPlanning.srv]: This is also taken from the original message definition. `candidate_grasps` sounds better to me too.

File: srv/GraspPlanning.srv [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+# the planning group used
+string group_name
+
+# the object to be grasped
+CollisionObject target
+
+# the name that the support surface (e.g. table) has in the collision map
+# can be left empty if no name is available
+string collision_support_surface_name
```
Comment (mikeferguson) [inline srv/GraspPlanning.srv]: while creating this message, should we make this an array? (Or maybe even a CollisionObject array?) This always seemed a bit underspecified to me, and I know I've at least once ran into the issue that an object is sitting half on one support surface and half on another.

**Comment by mikeferguson** [comment]
+1 to message consolidation, I'd suggest we give this at least 48 hours for review, since there might be additional good feedback on how make this message great again.

File: srv/GraspPlanning.srv (hunk 5 above)
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: +1

**Comment by davetcoleman** [comment]
Did I hear "Make MoveIt great again?"

File: srv/GraspPlanning.srv [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+# an optional list of grasps to be evaluated by the planner
+Grasp[] grasps_to_evaluate
+
+# an optional list of obstacles that we have semantic information about
+# and that can be moved in the course of grasping
+CollisionObject[] movable_obstacles
```
Comment (mikeferguson) [inline srv/GraspPlanning.srv]: I think the idea is that moveable_obstacles can actually be shifted by colliding with the gripper -- I'm pretty sure that none of our current planners can actually do anything with this, but there was some work going on at willow garage to do "push-based" manipulation and grasping, where the planner could figure out how to nudge things out of the way in order to get to the grasp and get a desired grasp.

Comment (jrgnicho) [inline srv/GraspPlanning.srv]: Right. I think the  potential usage you described is already encompassed in the __allowed_touch_objects__ entry in the [Grasp.msg](http://docs.ros.org/api/moveit_msgs/html/msg/Grasp.html) messages in this service definition; in fact they even have the same description.  I think we should avoid having entries with similar purpose as that could make the interface confusing.

Comment (mikeferguson) [inline srv/GraspPlanning.srv]: How does one specify the list of allowed touch objects as an input to the service though, if they have no initial estimate of grasps? The "grasps_to_evaluate"/"candidate_grasps" field is optional, and I presume most users probably won't be specifying it, but if we remove the moveable objects, then we have to pass at least one candidate grasp that contains only the allowed_touch_objects, but is not actually a candidate? That seems pretty unclean, and potentially leading to the sort of headaches we currently have with "is_diff" in the planning scene.

Comment (jrgnicho) [inline srv/GraspPlanning.srv]: @mikeferguson that's a good point.

File: srv/GraspPlanning.srv (hunk 5 above)
Comment (v4hn) [inline srv/GraspPlanning.srv]: At the moment, the MoveGroupInterface only supports setting *one* support surface and the `Pickup` and `Place` actions do too.
If we change this here, we should extend it at all other places in the kinetic branch. (Let's do that in a separate request)
I'm not sure this is useful in practice, but as at least three of you agree, let's change it.

**Comment by Jntzko** [comment]
I changed the service according to your feedback except for the "target object".

Also we've change the name for the collision surfaces. This fits better with the new array.

File: srv/GraspPlanning.srv (hunk 5 above)
Comment (davetcoleman) [inline srv/GraspPlanning.srv]: Or for now just add a warning if someone tries to pass in more than one supporting surface. ROS messages are just much harder/annoying to change after the fact, so they should be general enough in the first run

**Comment by v4hn** [comment]
ok, I'll merge this next Tuesday if nobody requests additional changes until then.

**Comment by v4hn** [comment]
ok, merging. I'll also pick this change to jade and kinetic

File: srv/GraspPlanning.srv (hunk 3 above)
Comment (rhaschke) [inline srv/GraspPlanning.srv]: +1 for the short name `target`.

**Comment by 130s** [comment]
> Did I hear "Make MoveIt great again?"

I might have heard "MoveIt! is great because it's good".

**Pull request merged**